packages/grid/_modules_/grid/hooks/features/useComponents.tsx

```diff
@@ -31,35 +36,23 @@ export const useComponents = (
     [pagination, rows, columns, options, apiRef, gridRootRef],
   );
 
-  const headerComponent = React.useMemo(
-    () =>
-      componentOverrides?.header
-        ? React.createElement(componentOverrides.header, componentProps)
-        : null,
-    [componentOverrides, componentProps],
-  );
-
-  const footerComponent = React.useMemo(
-    () =>
-      componentOverrides?.footer
-        ? React.createElement(componentOverrides.footer, componentProps)
-        : null,
-    [componentOverrides, componentProps],
-  );
+  const headerComponent = componentOverrides?.Header
+    ? React.createElement(componentOverrides.Header, componentProps)
+    : options.showToolbar && <DefaultToolbar />;
```

Comment:
I think this should follow the same login as the `Pagination` component - if the `options.showToolbar=false` then the custom header component is also hidden [#821](https://github.com/mui-org/material-ui-x/issues/821#issuecomment-756695797)

```suggestion
  const headerComponent = componentOverrides?.Header
    ? options.showToolbar && React.createElement(componentOverrides.Header, componentProps)
    : options.showToolbar && <DefaultToolbar />;
```

Reply:
I find this part confusing. How about we remove the `showToolbar` prop with #726?

Reply:
I agree It won't be needed anymore after [#726](https://github.com/mui-org/material-ui-x/issues/726) is done but shouldn't we keep it for now?

Reply:
Agree, I think that we should keep `showToolbar` as long as there aren't any better alternatives developers can use. Basically, we always need a migration path to solve a giving problem. 

I see two issues with the diff, If we introduce it, 1. we we make the BC even greater once we drop `showToolbar` 2. The toolbar isn't the header, I don't understand the logical jump between I don't want a toolbar therefor I don't want a header. In my understanding, these are two different notions.

As far as I understand it, the main value of the Header customization slot, and by a large margin, is to be inside the context providers of the data grid. It can unlock interesting APIs in the future.